addons/web/static/src/search/control_panel/control_panel.js

```diff
@@ -233,4 +335,204 @@ export class ControlPanel extends Component {
+            confirm: async () => await this._deleteTopBarAction(action),
+            cancel: () => {},
+        };
+        this.dialogService.add(ConfirmationDialog, dialogProps);
+    }
+
+    /**
+     * @param {TopBarAction} action
+     */
+    async _deleteTopBarAction(action) {
+        const actionIdStr = action.id.toString();
+        if (this.state.visibleTopBarActions[actionIdStr]) {
+            delete this.state.visibleTopBarActions[actionIdStr];
+        }
+        localStorage.setItem(
+            this.localStorageKeyActionIdResId,
+            JSON.stringify(this.state.visibleTopBarActions)
+        );
+        this.state.topBarActions = this.state.topBarActions.filter((el) => el.id !== action.id);
+        await this.orm.call("ir.actions.topbar", "unlink", [action.id]);
+        if (action.id === this.state.currentTopBarAction?.id) {
+            const context = {
+                ...makeContext([action.context]),
+                active_id: this.env.searchModel?.globalContext?.active_id,
+                active_model: this.env.searchModel?.globalContext?.active_model,
+                fromParentAction: true,
+            };
+            this.actionService.doAction(action.parent_action_id[0], {
+                additionalContext: context,
+                stackPosition: "replaceCurrentAction",
+            });
+        }
+    }
+
+    /**
+     * @param {TopBarAction} action
+     */
+    async onTopBarActionClick(action) {
+        this.env.config.topBarActions = this.state.topBarActions;
+        const context = {
+            ...makeContext([action.context]),
+            active_id: this.env.searchModel?.globalContext?.active_id,
+            active_model: this.env.searchModel?.globalContext?.active_model,
+            currentTopbarActionId: action.id,
+            parentActionChildren: this.env.config.topBarActions,
+            fromParentAction: true,
+        };
+        this.actionService.doActionButton({
+            type: action.python_action ? "object" : "action",
+            resId: this.env.searchModel?.globalContext.active_id,
+            name: action.python_action || action.action_id[0] || action.action_id,
+            resModel: action.res_model,
+            context: context,
+            stackPosition: this.env.config.fromParentAction ? "replaceCurrentAction" : "",
+            viewType: action.default_view_mode,
+        });
+    }
+
+    /**
+     * @param {Object} params
+     * @param {HTMLElement} params.element
+     */
+    _sortTopBarActionStart({ element, addClass }) {
+        addClass(element, "o_dragged_topBar_action");
+    }
+
+    /**
+     * @param {Object} params
+     * @param {HTMLElement} params.element
+     * @param {HTMLElement} params.previous
+     */
+    async _sortTopBarActionDrop({ element, previous }) {
+        const order = this.state.topBarActions.map((el) => el.id);
+        const elementId = Number(element.dataset.id);
+        const elementIndex = order.indexOf(elementId);
```

Comment:
```suggestion
        const elementIndex = this.state.topBarActions.findIndex(({ id }) => Number(element.dataset.id) === id);
```
And work with `this.state.topBarActions` instead of `order`?

Reply:
I don't think that is the best idea as I use splice method multiple times to compute the order, and that method is changing the array. So for the space added with the computation of order, I think that is worth it, but let me know if I am mistaken

Reply:
I think you're right